File: Algo/Leetcode/057InsertInterval.cpp
// 3 stage
class Solution {
public:
    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
        vector<vector<int>> res;
        int n = intervals.size();
        int i = 0;
        while(i < n && intervals[i][1] < newInterval[0]){
            res.push_back(intervals[i]);
            i++;
        }

        int min_interval = newInterval[0];
        int max_interval = newInterval[1];
        while(i < n && intervals[i][0] <= newInterval[1]){ //如果相等 仍可合并
            min_interval = min(min_interval, intervals[i][0]);
            max_interval = max(max_interval, intervals[i][1]);
            i++;
        }
        res.push_back(vector<int>{min_interval, max_interval});

        while(i<n){
            res.push_back(intervals[i]);
            i++;
        }

        return res;
    }
};

File: Algo/Leetcode/057InsertInterval.py
class Solution(object):
    def insert(self, intervals, newInterval):
        """
        :type intervals: List[List[int]]
        :type newInterval: List[int]
        :rtype: List[List[int]]
        """
        n = len(intervals)
        res = []
        if n == 0:
            return [newInterval]

        i = 0
        while i < n and intervals[i][1] < newInterval[0]:  # strictly larger
            res.append(intervals[i])
            i += 1

        start_t = newInterval[0]
        end_t = newInterval[1]
        while i <= n - 1 and intervals[i][0] <= newInterval[1]:  # consider the middle ones
            start_t = min(intervals[i][0], start_t)
            end_t = max(intervals[i][1], end_t)
            i += 1

        res.append([start_t, end_t])

        while i < n:
            res.append(intervals[i])
            i += 1

        return res
